support/historyarchive/log.go

```
@@ -0,0 +1,134 @@
+		} else {
+			changedBuckets += "_"
+		}
+		if prevHas.CurrentBuckets[i].Snap != b.Snap {
+			nChangedBuckets += 1
+			changedBuckets += "#"
+			nChangedBytes += arch.GetBucketSize(MustDecodeHash(b.Snap))
+		} else {
+			changedBuckets += "_"
+		}
+	}
+	return changedBuckets, nChangedBuckets, nChangedBytes
+}
+
+func (arch *Archive) GetBucketSize(hash Hash) int64 {
+	sz, err := arch.backend.Size(arch.GetBucketPathForHash(hash))
+	if err != nil {
+		panic(err)
+	}
+	return sz
+}
+
+func (arch *Archive) GetLedgerHeaderHistoryEntries(chk uint32) []xdr.LedgerHeaderHistoryEntry {
+	path := CategoryCheckpointPath("ledger", chk)
+	rdr, err := arch.GetXdrStream(path)
+	if err != nil {
+		panic(err)
+	}
+	defer rdr.Close()
+	var lhes []xdr.LedgerHeaderHistoryEntry
+	for {
+		lhe := xdr.LedgerHeaderHistoryEntry{}
+		if err = rdr.ReadOne(&lhe); err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				panic(err)
+			}
+		}
+		lhes = append(lhes, lhe)
+	}
+	return lhes
+}
+
+func (arch *Archive) GetTransactionHistoryEntries(chk uint32) []xdr.TransactionHistoryEntry {
+	path := CategoryCheckpointPath("transactions", chk)
+	rdr, err := arch.GetXdrStream(path)
+	if err != nil {
+		panic(err)
+	}
+	defer rdr.Close()
+	var thes []xdr.TransactionHistoryEntry
+	for {
+		the := xdr.TransactionHistoryEntry{}
+		if err = rdr.ReadOne(&the); err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				panic(err)
+			}
+		}
+		thes = append(thes, the)
+	}
+	return thes
+}
+
+func (arch *Archive) Log(opts *CommandOptions) error {
+	state, e := arch.GetRootHAS()
+	if e != nil {
+		return e
+	}
+	opts.Range = opts.Range.clamp(state.Range())
+
+	log.SetFlags(0)
+	log.Printf("Log of checkpoint files in range: %s", opts.Range)
+	log.Printf("\n")
+	log.Printf("%10s | %10s | %20s | %5s | %s",
+		"ledger", "hex", "close time", "txs", "buckets changed")
+
+	prevHas, err := arch.GetCheckpointHAS(PrevCheckpoint(opts.Range.Low))
+	if err != nil {
+		panic(err)
```

I found it surprising that this method both returns errors and panics. As a user of the method I would normally assume that if returning errors, then all errors are my responsibility as the caller. Do you feel strongly that you need to panic here rather than returning to the caller?

---

Oh, yeah, that's a mistake. Fixed.